File: 05_Recursion/Recursive_Factorial/main.go
// Recursion

package main

import "fmt"

func fun(n int) int {
	if n == 0 { //0!=1
		return 1
	}
	if n == 1 {
		return 1
	}
	return fun(n-1) * n
}

func main() {
	a := 4
	fmt.Println(fun(a))
}

// //Iteration

// package main

// import "fmt"

// func fun(v int) int {
// 	if v == 0 { //0!=1
// 		return 1
// 	}
// 	if v == 1 {
// 		return 1
// 	}

// 	fact := 1
// 	for i := 1; i <= v; i++ {
// 		fact = fact * i
// 	}
// 	return fact
// }

// func main() {
// 	a := 4
// 	fmt.Println(fun(a))
// }

File: 05_Recursion/Recursive_Factorial/main.cpp
// #include<iostream>
// using namespace std;

// int fun (int n) {
//     if (n == 0) //0!=1
//     return 1;
    
//     if (n == 1)
//     return 1;
    
//     return fun(n - 1) * n;    
// }

// int main (){
//     int a = 4;
//     printf ("%d\n", fun(a));
//     return 0;
// }

//Iteration
#include<iostream>
using namespace std;

int fun (int v) {
    if (v == 0) //0!=1
    return 1;
    
    if (v == 1)
    return 1;

    int v, fact;
    fact = 1;
    for (int i = 1; i <= v; i++) {
       fact = fact * i;
    } 
    return fact;
}

int main (){
    int a = 4;
    printf ("%d\n", fun(a));
    return 0;
}

